### PR TITLE
feat(types): add generics types for associations

### PR DESCRIPTION
**Pull request by Kilimangaro**
### Description of change

Added generics types (default to "typeof Model) to all associations class definitions:

- HasOne
- HasMany
- BelongsTo
- BelongsToMany

These associations can now be defined as follow

```typescript
// ...
public static associations: {
  group: BelongsTo<User, UserGroup>;
};
// ...

/**
 * userType is now a "typeof User" instead of "typeof Model" 
 */
const userType: typeof User = User.associations.group.source;
```
Also, associations methods now returns the typed associations

```typescript
/**
 * Group is now a "BelongsTo<User, UserGroup>" instead of "BelongsTo" 
 */
export const Group = User.belongsTo(UserGroup, { as: 'group', foreignKey: 'groupId' });
```

### CODE REVIEW

**Comment by sushantdhiman** [comment]
LGTM, @SimonSchick please review

File: types/lib/associations/base.d.ts [hunk 1]
```diff
@@ -1,9 +1,9 @@
 import { ColumnOptions, Model } from '../model';
 
-export abstract class Association {
+export abstract class Association<S = typeof Model, T = typeof Model> {
```
Comment (SimonSchick) [inline types/lib/associations/base.d.ts]: Can you please add an export to `model.d.ts` that's called `ModelType = typeof Model` and use that instead?

File: types/lib/model.d.ts [hunk 2]
```diff
@@ -2235,7 +2235,10 @@ export abstract class Model<T = any, T2 = any> extends Hooks {
    * @param target The model that will be associated with hasOne relationship
    * @param options Options for the association
    */
-  public static hasOne(target: typeof Model, options?: HasOneOptions): HasOne;
+  public static hasOne<M, T extends typeof Model>(
+    this: M,
```
Comment (SimonSchick) [inline types/lib/model.d.ts]: Should probably be `{ new (): M } & typeof Model` but feel free to create a generic such as `ModelCtor<M> = { new (): M } & typeof Mode` to clean this up a bit.

File: types/lib/model.d.ts [hunk 3]
```diff
@@ -2300,7 +2306,10 @@ export abstract class Model<T = any, T2 = any> extends Hooks {
    * @param target The model that will be associated with hasOne relationship
    * @param options Options for the association
    */
-  public static hasMany(target: typeof Model, options?: HasManyOptions): HasMany;
+  public static hasMany<M, T extends typeof Model>(
+    this: M, 
+    target: T, options?: HasManyOptions
```
Comment (SimonSchick) [inline types/lib/model.d.ts]: Nit: Please be consistent about newlines

File: types/lib/associations/base.d.ts (hunk 1 above)
Comment (Kilimangaro) [reply types/lib/associations/base.d.ts]: done

File: types/lib/model.d.ts (hunk 2 above)
Comment (Kilimangaro) [reply types/lib/model.d.ts]: Added  export type "ModelCtor"

File: types/lib/model.d.ts (hunk 3 above)
Comment (Kilimangaro) [reply types/lib/model.d.ts]: done.

It could be nice to have some prettier config

File: types/test/models/User.ts [hunk 4]
```diff
@@ -68,7 +71,11 @@ User.afterFind((users, options) => {
 User.addHook('beforeFind', 'test', (options: FindOptions) => {
   return undefined;
 });
+
 // associate
 // it is important to import _after_ the model above is already exported so the circular reference works.
 import { UserGroup } from './UserGroup';
 export const Group = User.belongsTo(UserGroup, { as: 'group', foreignKey: 'groupId' });
+
+// const userType: ModelCtor<User> = User.associations.group.source;
```
Comment (sushantdhiman) [inline types/test/models/User.ts]: Why are these lines commented?

Comment (Kilimangaro) [reply types/test/models/User.ts]: my mistake... fixed that in last commit

**Pull request merged**